onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -77,19 +90,38 @@ CastToString(const SrcType& input, std::string& output) {
       output = "INF";
     }
   } else {
-    // setprecision to 8 to match numpy default behavior
-    std::ostringstream convert;
-    convert << std::setprecision(8) << input;
-    output = convert.str();
+    // set precision to 8 to match numpy default behavior
+    constexpr const char* format = "%.8g";
+    const double value = static_cast<double>(input);
+
+    auto snprintf_result = std::snprintf(nullptr, 0, format, value);
+    ORT_ENFORCE(snprintf_result > 0, "Failed to determine required snprintf() buffer length.");
+
+    // buffer for string and trailing '\0'
+    InlinedBuffer<char, 32> buffer(gsl::narrow<size_t>(snprintf_result + 1));
+
+    snprintf_result = std::snprintf(buffer.data(), buffer.size(), format, value);
+    ORT_ENFORCE(
+        snprintf_result > 0 && gsl::narrow_cast<size_t>(snprintf_result) == buffer.size() - 1,
+        "Failed to write value with snprintf().");
+
```

Comment:
Prefer returning a status in code that runs during model execution. In a minimal build that will gracefully return an error (vs. ORT_ENFORCE which will call abort() if exceptions are disabled).

Reply:
ok, good to know. however i think this is sufficiently bad to warrant termination

---
In reply to: [569864399](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569864399) [](ancestors = 569864399)

Reply:
Either way the request will terminate as we essentially exit as soon as the status indicates failure.  If you return a Status that will come back to the user gracefully with an error message in all builds. If you throw, it will do the same unless exceptions are disabled, in which case it's a hard abort which is much harder to debug.

---
In reply to: [569930622](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569930622) [](ancestors = 569930622,569864399)

Reply:
ok. the error message still gets printed though, right?
let me know if you have a strong preference for Status in this particular case. it'll require some plumbing to propagate it. i think these snprintf failures are unexpected - i.e., more like an assertion

---
In reply to: [569957335](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569957335) [](ancestors = 569957335,569930622,569864399)

Reply:
trying out Status

---
In reply to: [570436510](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r570436510) [](ancestors = 570436510,569957335,569930622,569864399)